File: rd-kt/rd-text/src/main/kotlin/com/jetbrains/rd/rdtext/impl/ot/OtFramework.kt
package com.jetbrains.rd.rdtext.impl.ot

import com.jetbrains.rd.util.collections.ImmutableStack
import com.jetbrains.rd.util.collections.tail
import com.jetbrains.rd.util.collections.toImmutableStack

private fun <T> MutableList<T>.append(el: T): MutableList<T> = this.apply { add(el) }

/**
 * Not commutative composition function.
 *
 * Notes: Use it if you want to merge a lot not optimized changes into one.
 * But all of them must have the same logical timestamp of IDE state (for example they must be generated during the same write action).
 */
fun compose(o1: OtOperation, o2: OtOperation): OtOperation {

    val after = o1.documentLengthAfter()
    val before = o2.documentLengthBefore()
    require(after == before, { "length after o1($after) != length before o2($before)" })
    require(o1.origin == o2.origin)
    require(o1.timestamp == o2.timestamp)
    require(o1.kind == o2.kind)

    tailrec fun compose0(acc: MutableList<OtChange>, ops1: ImmutableStack<OtChange>, ops2: ImmutableStack<OtChange>) {
        val op1 = ops1.peek()
        val op2 = ops2.peek()

        val tail1 = ops1.tail()
        val tail2 = ops2.tail()
        when {
            op1 == null && op2 == null -> Unit
            op1 != null && op2 == null -> compose0(acc.append(op1), tail1, tail2)
            op1 == null && op2 != null -> compose0(acc.append(op2), tail1, tail2)
            op1 is DeleteText && op2 is InsertText && op1.text == op2.text -> compose0(acc.append(Retain(op1.text.length)), tail1, tail2)
            op2 is InsertText -> compose0(acc.append(op2), ops1, tail2)
            op1 is DeleteText -> compose0(acc.append(op1), tail1, ops2)
            op1 is Retain && op2 is Retain -> {
                val offset1 = op1.offset
                val offset2 = op2.offset
                when {
                    offset1 > offset2 -> compose0(acc.append(op2), tail1.push(Retain(offset1 - offset2)), tail2)
                    offset1 == offset2 -> compose0(acc.append(op1), tail1, tail2)
                    offset1 < offset2 -> compose0(acc.append(op1), tail1, tail2.push(Retain(offset2 - offset1)))
                    else -> throw IllegalArgumentException("op1 is Retain && op2 is Retain")
                }
            }
            op1 is InsertText && op2 is DeleteText -> {
                val text1 = op1.text
                val text2 = op2.text
                when {
                    text1.length > text2.length -> {
                        require(text1.startsWith(text2))
                        compose0(acc, tail1.push(InsertText(text1.substring(text2.length))), tail2)
                    }
                    text1.length == text2.length -> {
                        require(text1 == text2)
                        compose0(acc, tail1, tail2)
                    }
                    text1.length < text2.length -> {
                        require(text2.startsWith(text1))
                        compose0(acc, tail1, tail2.push(DeleteText(text2.substring(text1.length))))
                    }
                    else -> throw IllegalArgumentException("op1 is InsertText && op2 is DeleteText")
                }
            }
            op1 is InsertText && op2 is Retain -> {
                val text1 = op1.text
                val offset2 = op2.offset
                when {
                    text1.length > offset2 -> compose0(acc.append(InsertText(text1.substring(0, offset2))),
                            tail1.push(InsertText(text1.substring(offset2))), tail2)
                    text1.length == offset2 -> compose0(acc.append(op1), tail1, tail2)
                    text1.length < offset2 -> compose0(acc.append(op1), tail1, tail2.push(Retain(offset2 - text1.length)))
                    else -> throw IllegalArgumentException("op1 is InsertText && op2 is DeleteText")
                }
            }
            op1 is Retain && op2 is DeleteText -> {
                val offset1 = op1.offset
                val text2 = op2.text
                when {
                    offset1 > text2.length -> compose0(acc.append(op2), tail1.push(Retain(offset1 - text2.length)), tail2)
                    offset1 == text2.length -> compose0(acc.append(op2), tail1, tail2)
                    offset1 < text2.length -> compose0(acc.append(DeleteText(text2.substring(0, offset1))),
                            tail1, tail2.push(DeleteText(text2.substring(offset1))))
                    else -> throw IllegalArgumentException("op1 is InsertText && op2 is DeleteText")
                }
            }
            else -> throw IllegalArgumentException("Not matched pair: ($op1, $op2)")
        }
    }

    val acc = mutableListOf<OtChange>()
    compose0(acc, o1.changes.toImmutableStack(), o2.changes.toImmutableStack())

    return OtOperation(acc, o1.origin, o1.timestamp, o1.kind)
}



data class OtTransformResult(val newLocalDiff: OtOperation, val localizedApplyToDocument: OtOperation)


// Ressel’s transformation function
fun transform(localDiff: OtOperation, remoteApplyToDocument: OtOperation): OtTransformResult {

    require(localDiff.documentLengthBefore() == remoteApplyToDocument.documentLengthBefore())
    require(localDiff.origin != remoteApplyToDocument.origin)
    require(localDiff.kind == OtOperationKind.Normal && remoteApplyToDocument.kind == OtOperationKind.Normal)

    tailrec fun transform0(resOp1: MutableList<OtChange>, resOp2: MutableList<OtChange>, ops1: ImmutableStack<OtChange>, ops2: ImmutableStack<OtChange>) {
        val op1: OtChange? = ops1.peek()
        val op2: OtChange? = ops2.peek()

        if (op1 == null && op2 == null) return

        val tail1 = ops1.tail()
        val tail2 = ops2.tail()

        when {
            op1 == null && op2 != null -> {
                val offset = op2.getTextLengthAfter()
                if (offset > 0)
                    transform0(resOp1.append(Retain(offset)), resOp2.append(op2), tail1, tail2)
                else
                    transform0(resOp1, resOp2.append(op2), tail1, tail2)
            }
            op1 != null && op2 == null -> {
                val offset = op1.getTextLengthAfter()
                if (offset > 0)
                    transform0(resOp1.append(op1), resOp2.append(Retain(offset)), tail1, tail2)
                else
                    transform0(resOp1.append(op1), resOp2, tail1, tail2)
            }
            op1 is InsertText && op2 is InsertText -> {
                if (localDiff.origin < remoteApplyToDocument.origin)
                    transform0(resOp1.append(op1), resOp2.append(Retain(op1.text.length)), tail1, ops2)
                else
                    transform0(resOp1.append(Retain(op2.text.length)), resOp2.append(op2), ops1, tail2)
            }
            op1 is InsertText -> transform0(resOp1.append(op1), resOp2.append(Retain(op1.text.length)), tail1, ops2)
            op2 is InsertText -> transform0(resOp1.append(Retain(op2.text.length)), resOp2.append(op2), ops1, tail2)
            op1 is Retain && op2 is Retain -> {
                val offset1 = op1.offset
                val offset2 = op2.offset
                when {
                    offset1 > offset2 -> transform0(resOp1.append(op2), resOp2.append(op2), tail1.push(Retain(offset1 - offset2)), tail2)
                    offset1 == offset2 -> transform0(resOp1.append(op1), resOp2.append(op1), tail1, tail2)
                    offset1 < offset2 -> transform0(resOp1.append(op1), resOp2.append(op1), tail1, tail2.push(Retain(offset2 - offset1)))
                }
            }
            op1 is DeleteText && op2 is DeleteText -> {
                val text1 = op1.text
                val text2 = op2.text
                when {
                    text1.length > text2.length -> transform0(resOp1, resOp2, tail1.push(DeleteText(text1.substring(text2.length))), tail2)
                    text1.length == text2.length -> {
                        require(text1 == text2)
                        transform0(resOp1, resOp2, tail1, tail2)
                    }
                    text1.length < text2.length -> transform0(resOp1, resOp2, tail1, tail2.push(DeleteText(text2.substring(text1.length))))
                }
            }
            op1 is DeleteText && op2 is Retain -> {
                val text1 = op1.text
                val offset2 = op2.offset
                when {
                    text1.length > offset2 -> transform0(resOp1.append(DeleteText(text1.substring(0, offset2))), resOp2,
                            tail1.push(DeleteText(text1.substring(offset2))), tail2)
                    text1.length == offset2 -> transform0(resOp1.append(op1), resOp2, tail1, tail2)
                    text1.length < offset2 -> transform0(resOp1.append(op1), resOp2, tail1, tail2.push(Retain(offset2 - text1.length)))
                }
            }
            op1 is Retain && op2 is DeleteText -> {
                val offset1 = op1.offset
                val text2 = op2.text
                when {
                    offset1 > text2.length -> transform0(resOp1, resOp2.append(op2), tail1.push(Retain(offset1 - text2.length)), tail2)
                    offset1 == text2.length -> transform0(resOp1, resOp2.append(op2), tail1, tail2)
                    offset1 < text2.length -> transform0(resOp1, resOp2.append(DeleteText(text2.substring(0, offset1))),
                            tail1, tail2.push(DeleteText(text2.substring(offset1))))
                }
            }
        }
    }

    val newLocalDiff = ArrayList<OtChange>()
    val localizedApplyToDocument = ArrayList<OtChange>()
    transform0(newLocalDiff, localizedApplyToDocument, localDiff.changes.toImmutableStack(), remoteApplyToDocument.changes.toImmutableStack())

    return OtTransformResult(
            OtOperation(newLocalDiff, localDiff.origin, localDiff.timestamp, OtOperationKind.Normal),
            OtOperation(localizedApplyToDocument, remoteApplyToDocument.origin, remoteApplyToDocument.timestamp, OtOperationKind.Normal)
    )
}

File: rd-net/RdFramework/Text/Impl/Ot/OtFramework.cs
using System;
using System.Collections.Generic;
using System.Linq;
using JetBrains.Diagnostics;

namespace JetBrains.Rd.Text.Impl.Ot
{
  public struct OtTransformResult
  {
    public readonly OtOperation NewLocalDiff;
    public readonly OtOperation LocalizedApplyToDocument;

    public OtTransformResult(OtOperation newLocalDiff, OtOperation localizedApplyToDocument)
    {
      NewLocalDiff = newLocalDiff;
      LocalizedApplyToDocument = localizedApplyToDocument;
    }
  }

  public static class OtFramework
  {
/*
    public static OtOperation Compose(this OtOperation o1, OtOperation o2)
    {
      var after = o1.DocumentLengthAfter();
      var before = o2.DocumentLengthBefore();

      Assertion.Assert(after == before, "after == before");
      Assertion.Assert(o1.Origin == o2.Origin, "o1.Role == o2.Role");
      Assertion.Assert(o1.Timestamp == o2.Timestamp, "o1.Timestamp == o2.Timestamp");
      Assertion.Assert(o1.Kind == o2.Kind, "o1.Kind == o2.Kind");


      var ops1 = new Stack<OtChange>(Enumerable.Reverse(o1.Changes));
      var ops2 = new Stack<OtChange>(Enumerable.Reverse(o2.Changes));
      var acc = new List<OtChange>();

      while (true)
      {
        var op1 = ops1.Count != 0 ? ops1.Peek() : null;
        var op2 = ops2.Count != 0 ? ops2.Peek() : null;

        if (op1 == null && op2 == null)
          break;

        if (op1 != null && op2 == null)
        {
          acc.Add(op1);
          ops1.Pop();
          continue;
        }

        if (op1 == null && op2 != null)
        {
          acc.Add(op2);
          ops2.Pop();
          continue;
        }

        switch (op1)
        {
          case DeleteText d1 when op2 is InsertText d2 && d1.Text == d2.Text:
          {
            var length = d1.Text.Length;
            acc.Add(new Retain(length));
            ops1.Pop();
            ops2.Pop();
            break;
          }
          case OtChange _ when op2 is InsertText:
          {
            acc.Add(op2);
            ops2.Pop();
            break;
          }
          case DeleteText _ when op2 != null:
          {
            acc.Add(op1);
            ops1.Pop();
            break;
          }
          case Retain r1 when op2 is Retain r2:
          {
            var offset1 = r1.Offset;
            var offset2 = r2.Offset;
            if (offset1 > offset2)
            {
              acc.Add(op2);
              ops1.Pop();
              ops2.Pop();
              ops1.Push(new Retain(offset1 - offset2));
            }
            else if (offset1 == offset2)
            {
              acc.Add(op1);
              ops1.Pop();
              ops2.Pop();
            }
            else if (offset1 < offset2)
            {
              acc.Add(op1);
              ops1.Pop();
              ops2.Pop();
              ops2.Push(new Retain(offset2 - offset1));
            }

            break;
          }
          case InsertText i1 when op2 is DeleteText d2:
          {
            var text1 = i1.Text;
            var text2 = d2.Text;
            if (text1.Length > text2.Length)
            {
              Assertion.Assert(text1.StartsWith(text2), "text1.StartsWith(text2)");
              ops1.Pop();
              ops2.Pop();
              ops1.Push(new InsertText(text1.Substring(text2.Length)));
            }
            else if (text1.Length == text2.Length)
            {
              Assertion.Assert(text1 == text2, "text1 == text2");
              ops1.Pop();
              ops2.Pop();
            }
            else if (text1.Length < text2.Length)
            {
              Assertion.Assert(text2.StartsWith(text1), "text2.StartsWith(text1)");
              ops1.Pop();
              ops2.Pop();
              ops2.Push(new DeleteText(text2.Substring(text1.Length)));
            }

            break;
          }
          case InsertText i1 when op2 is Retain r2:
          {
            var text1 = i1.Text;
            var offset2 = r2.Offset;
            if (text1.Length > offset2)
            {
              acc.Add(new InsertText(text1.Substring(0, offset2)));
              ops1.Pop();
              ops2.Pop();
              ops1.Push(new InsertText(text1.Substring(offset2)));
            }
            else if (text1.Length == offset2)
            {
              acc.Add(op1);
              ops1.Pop();
              ops2.Pop();
            }
            else if (text1.Length < offset2)
            {
              acc.Add(op1);
              ops1.Pop();
              ops2.Pop();
              ops2.Push(new Retain(offset2 - text1.Length));
            }

            break;
          }
          case Retain r1 when op2 is DeleteText d2:
          {
            var offset1 = r1.Offset;
            var text2 = d2.Text;
            if (offset1 > text2.Length)
            {
              acc.Add(op2);
              ops1.Pop();
              ops2.Pop();
              ops1.Push(new Retain(offset1 - text2.Length));
            }
            else if (offset1 == text2.Length)
            {
              acc.Add(op2);
              ops1.Pop();
              ops2.Pop();
            }
            else if (offset1 < text2.Length)
            {
              acc.Add(new DeleteText(text2.Substring(0, offset1)));
              ops1.Pop();
              ops2.Pop();
              ops2.Push(new DeleteText(text2.Substring(offset1)));
            }

            break;
          }
          default: throw new ArgumentOutOfRangeException($"Not matched pair: (op1 = {op1}, op2 = {op2})");
        }
      }

      return new OtOperation(acc, o1.Origin, o1.Timestamp, o1.Kind);
    }
*/

    // Ressel’s transformation function
    public static OtTransformResult Transform(OtOperation localDiff, OtOperation remoteApplyToDocument)
    {
      Assertion.Assert(localDiff.DocumentLengthBefore() == remoteApplyToDocument.DocumentLengthBefore(),
        "localDiff.DocumentLengthBefore() == remoteApplyToDocument.DocumentLengthBefore()");
      Assertion.Assert(localDiff.Origin != remoteApplyToDocument.Origin, "localDiff.Role != remoteApplyToDocument.Role");
      Assertion.Assert(localDiff.Kind == OtOperationKind.Normal && remoteApplyToDocument.Kind == OtOperationKind.Normal,
        "localDiff.Kind == OtOperationKind.Normal && remoteApplyToDocument.Kind == OtOperationKind.Normal");

      var resOp1 = new List<OtChange>();
      var resOp2 = new List<OtChange>();
      var ops1 = new Stack<OtChange>(Enumerable.Reverse(localDiff.Changes));
      var ops2 = new Stack<OtChange>(Enumerable.Reverse(remoteApplyToDocument.Changes));

      while (true)
      {
        var op1 = ops1.Count != 0 ? ops1.Peek() : null;
        var op2 = ops2.Count != 0 ? ops2.Peek() : null;

        if (op1 == null && op2 == null)
          break;

        if (op1 == null && op2 != null)
        {
          var offset = op2.GetTextLengthAfter();
          if (offset > 0)
          {
            resOp1.Add(new Retain(offset));
            resOp2.Add(op2);
            ops2.Pop();
          }
          else
          {
            resOp2.Add(op2);
            ops2.Pop();
          }

          continue;
        }

        if (op1 != null && op2 == null)
        {
          var offset = op1.GetTextLengthAfter();
          if (offset > 0)
          {
            resOp1.Add(op1);
            resOp2.Add(new Retain(offset));
            ops1.Pop();
          }
          else
          {
            resOp2.Add(op1);
            ops1.Pop();
          }

          continue;
        }


        switch (op1)
        {
          case InsertText i1 when op2 is InsertText i2:
          {
            if (localDiff.Origin < remoteApplyToDocument.Origin)
            {
              resOp1.Add(i1);
              resOp2.Add(new Retain(i1.Text.Length));
              ops1.Pop();
            }
            else
            {
              resOp1.Add(new Retain(i2.Text.Length));
              resOp2.Add(op2);
              ops2.Pop();
            }

            break;
          }
          case InsertText i1:
          {
            resOp1.Add(i1);
            resOp2.Add(new Retain(i1.Text.Length));
            ops1.Pop();
            break;
          }
          case var _ when op2 is InsertText i2:
          {
            resOp1.Add(new Retain(i2.Text.Length));
            resOp2.Add(i2);
            ops2.Pop();
            break;
          }
          case Retain r1 when op2 is Retain r2:
          {
            var offset1 = r1.Offset;
            var offset2 = r2.Offset;
            if (offset1 > offset2)
            {
              resOp1.Add(op2);
              resOp2.Add(op2);
              ops1.Pop();
              ops2.Pop();
              ops1.Push(new Retain(offset1 - offset2));
            }
            else if (offset1 == offset2)
            {
              resOp1.Add(op1);
              resOp2.Add(op1);
              ops1.Pop();
              ops2.Pop();
            }
            else if (offset1 < offset2)
            {
              resOp1.Add(op1);
              resOp2.Add(op1);
              ops1.Pop();
              ops2.Pop();
              ops2.Push(new Retain(offset2 - offset1));
            }

            break;
          }
          case DeleteText d1 when op2 is DeleteText d2:
          {
            var text1 = d1.Text;
            var text2 = d2.Text;
            if (text1.Length > text2.Length)
            {
              ops1.Pop();
              ops2.Pop();
              ops1.Push(new DeleteText(text1.Substring(text2.Length)));
            }
            else if (text1.Length == text2.Length)
            {
              Assertion.Assert(text1 == text2, "text1 == text2");
              ops1.Pop();
              ops2.Pop();
            }
            else if (text1.Length < text2.Length)
            {
              ops1.Pop();
              ops2.Pop();
              ops2.Push(new DeleteText(text2.Substring(text1.Length)));
            }

            break;
          }
          case DeleteText d1 when op2 is Retain r2:
          {
            var text1 = d1.Text;
            var offset2 = r2.Offset;
            if (text1.Length > offset2)
            {
              resOp1.Add(new DeleteText(text1.Substring(0, offset2)));
              ops1.Pop();
              ops2.Pop();
              ops1.Push(new DeleteText(text1.Substring(offset2)));
            }
            else if (text1.Length == offset2)
            {
              resOp1.Add(op1);
              ops1.Pop();
              ops2.Pop();
            }
            else if (text1.Length < offset2)
            {
              resOp1.Add(op1);
              ops1.Pop();
              ops2.Pop();
              ops2.Push(new Retain(offset2 - text1.Length));
            }

            break;
          }
          case Retain r1 when op2 is DeleteText r2:
          {
            var offset1 = r1.Offset;
            var text2 = r2.Text;
            if (offset1 > text2.Length)
            {
              resOp2.Add(op2);
              ops1.Pop();
              ops2.Pop();
              ops1.Push(new Retain(offset1 - text2.Length));
            }
            else if (offset1 == text2.Length)
            {
              resOp2.Add(op2);
              ops1.Pop();
              ops2.Pop();
            }
            else if (offset1 < text2.Length)
            {
              resOp2.Add(new DeleteText(text2.Substring(0, offset1)));
              ops1.Pop();
              ops2.Pop();
              ops2.Push(new DeleteText(text2.Substring(offset1)));
            }
            break;
          }
          default: throw new ArgumentOutOfRangeException($"Not matched pair: (op1 = {op1}, op2 = {op2})");
        }
      }

      return new OtTransformResult(
        new OtOperation(resOp1, localDiff.Origin, localDiff.Timestamp, OtOperationKind.Normal),
        new OtOperation(resOp2, remoteApplyToDocument.Origin, remoteApplyToDocument.Timestamp, OtOperationKind.Normal));
    }
  }
}